apache2/apache2_config.c

```diff
@@ -290,14 +297,11 @@ static int copy_rules(apr_pool_t *mp, msre_ruleset *parent_ruleset,
                       msre_ruleset *child_ruleset,
                       apr_array_header_t *exceptions_arr)
 {
+    assert(parent_ruleset != NULL);
+    assert(child_ruleset != NULL);
+    assert(exceptions_arr != NULL);
     int ret = 0;
 
-    if (parent_ruleset == NULL || child_ruleset == NULL ||
```

Comment:
I think leaving these checks and NULL pointer handling parts is not a good idea.

As we discussed an `assert()` function can have two results:
* if someone uses `-DNDEBUG` `CFLAG`, then it terminates the program if condition is false (this is available with `--enable-assertions` configure option)
* if `-DNDEBUG` is not used (default configure setting) then `assert()` does nothing

Now consider if any of checked variable is NULL (condition is false), `-DNDEBUG` is not used -> `assert()` does nothing. Example: `parent_ruleset` is NULL. `assert()` does nothing, there is no other check, and in the next line we pass a NULL pointer to the `copy_rules_phase()` function.

Same situation in case of each removed NULL ptr check.

Reply:
The goal of these assert() is to check conditions that can **never** happen. They are a watch dog that, once the development is complete, should be useless. This is why there's no "real" check in the normal code.
In this example, `parent_ruleset` cannot NULL; that's guaranteed by httpd (and our initialization). It cannot be NULL because of a run-time property. The only way it can be NULL is if the code is incorrect in its design (really design, not forgetting to check for NULL before calling the function). If it's incorrect in such a way, the whole behaviour of the application is screwed up. And, in almost all cases, even if we return correctly instead of referencing a NULL pointer (which is almost equivalent to an assert btw, except that logging is quite ugly), the code following the call will, at a certain moment, be in a totally inconsistent state. Because that inconsistent state may be something different than a NULL pointer, it's even worse because we could have memory corruption, which won't happen with a NULL pointer deferencing.

Reply:
> The goal of these assert() is to check conditions that can **never** happen.

Can we guarantee this at this moment?

> This is why there's no "real" check in the normal code.

I see, but until we can't guarantee the conditions, I think we can't leave the checks.

> In this example, `parent_ruleset` cannot NULL; 

`parent_ruleset` cannot. But I tried to review the functions where this function is called, and found this block:

https://github.com/owasp-modsecurity/ModSecurity/blob/v2/master/apache2/apache2_config.c#L429-L437

If I'm right, in this case the `rule_exceptions` will be NULL.

I think introduce `assert()` is a good direction, but at this moment we can't leave the regular checks. We just can't guarantee that every conditions are satisfied.

Reply:
I re-added all NULL check at run-time, with an error message on stderr (as it's a critical problem).